### PR TITLE
FSDP use _allgather_base and _reduce_scatter_base

### PR DESCRIPTION
**Pull request by liangluofb**
## What does this PR do?
Fixes # (issue).
We added support for flattened tensor allgather and reduce scatter to remove excessive memory copy. This patch makes FSDP use those functions.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

**Comment by anj-s** [comment]
Can you add a description to the PR and context about why we want to use private functions (If i remember correctly this prevents a second copy of tensors and instead uses a flattened tensor right from the start resulting in a speedup)? Thanks!

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (min-xu-ai) [inline fairscale/utils/reduce_scatter_bucketer.py:29]: is this going to work in different pytorch version? Perhaps need a version check at the top of the file and set a boolean and check that boolean to decide which API to use?

**Review by sshleifer — APPROVED** [review]
Nice! 
Interested in the speedup here if anyone has run benchmarks!
LGTM pending @min-xu-ai 's version comment, CI.

**Comment by anj-s** [comment]
> Nice!
> Interested in the speedup here if anyone has run benchmarks!
> LGTM pending @min-xu-ai 's version comment, CI.

We don't have regression benchmarks for FSDP which has been a longstanding [task](https://github.com/facebookresearch/fairscale/issues/677). @liangluofb  Can you share some numbers on what is the expected speedup and memory usage changes?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (myleott) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: Ideally we would use non-blocking GPU -> CPU transfer here. This was originally enabled, but I had to disable it in #496 because I found it was producing non-identical results.

I was not able to build a unit test to reproduce this problem, but it happens with the fairseq example here: https://github.com/pytorch/fairseq/tree/master/examples/fully_sharded_data_parallel#13b-params-on-1-v100-gpu-with-cpu-offloading

Basically, you should see identical results with non_blocking=True and non_blocking=False, but I didn't see identical results at the time. This suggests a missing stream synchronization or something similar.

Probably this will be fixed by an extra `torch.cuda.synchronize()` in fairseq's cpu_adam.py, which is being added here: https://github.com/pytorch/fairseq/pull/3650/files#diff-a7721b9e9db5b90f71bb01c359c13835fd98d30562e5af9127cb388d53ea65c7

We should confirm that combining these two changes (non_blocking=True + torch.cuda.synchronize() at in cpu_adam.py) will yield identical results.

Comment (liangluofb) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: Hi Myle,
I get identical loss for 8 GPUs 30 iterations,dummy_lm, with and without these changes, on gpt-13, full offloading and  checkpointing, cpu_adam (with that stream synchronization).

WO
```
I0628 140700.318 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3120.13 | wps 2669.2 | ups 0.02 | wpb 131072 | bsz 64 | num_updates 30 | lr 0.0001 | gnorm 3.806 | loss_scale 2 | train_wall 1726 | gb_free 16.7 | wall 1794
```           

With:
```
I0628 141512.471 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3120.13 | wps 2322.4 | ups 0.02 | wpb 131072 | bsz 64 | num_updates 30 | lr 0.0001 | gnorm 3.806 | loss_scale 2 | train_wall 1953 | gb_free 16.7 | wall 2022
```

As well as on a single GPU:

WO:
```
I0628 132805.584 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3119.93 | wps 125.1 | ups 0.01 | wpb 16384 | bsz 8 | num_updates 30 | lr 0.0001 | gnorm 3.387 | loss_scale 4 | train_wall 4286 | gb_free 16.7 | wall 4357
```

With:
```
I0628 133745.308 progress_bar.py:285] epoch 001 | loss 11.607 | ppl 3119.93 | wps 133 | ups 0.01 | wpb 16384 | bsz 8 | num_updates 30 | lr 0.0001 | gnorm 3.387 | loss_scale 4 | train_wall 4011 | gb_free 16.7 | wall 4084
```

**Comment by liangluofb** [comment]
> > Nice!
> > Interested in the speedup here if anyone has run benchmarks!
> > LGTM pending @min-xu-ai 's version comment, CI.
> 
> We don't have regression benchmarks for FSDP which has been a longstanding [task](https://github.com/facebookresearch/fairscale/issues/677). @liangluofb Can you share some numbers on what is the expected speedup and memory usage changes?

It varies a lot by models. The largest we see, on 128 GPUs for XLM-R XL, is 1.7x

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (myleott) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: Thanks, looks awesome!

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (liangluofb) [reply fairscale/utils/reduce_scatter_bucketer.py]: Hi! Thanks for pointing this out. Right now I think we can leave it for later?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: Does the WPS (words per second) from the logs above suggest that this slows down training?

### My interpretation
- I assumed "WO"=> "Without these changes" => master
```
N-gpu/master: 2669.2
N-gpu/branch: 2322.4 

1-gpu/master: 125.1 
1-gpu/branch: 133
```
=> this change introduces a 14% slowdown (2669/2332-1) for the N-GPU case and a 6% speedup for the 1 GPU case.

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (anj-s) [reply fairscale/utils/reduce_scatter_bucketer.py]: Sure, can we add a TODO so that we can look into it later?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (anj-s) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: Question related to the torch.cuda.synchronize() fix. Is this something we need to do in the step() function of any optimizer? Should we test this with non-fairseq code?

Comment (liangluofb) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: Nope, only if we are offloading to CPUs, because there is nothing that prevents CPU from getting ahead and not waiting for the GPUs to transfer the proper gradients.

Comment (liangluofb) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1318]: If the optimizer is on GPU, the stream semantics ensures the ordering (as long as the optimizer and the gradient are processed on the same stream)

**Comment by anj-s** [comment]
Friendly ping on this PR :) Once we address the last few comments and lint errors this should be good to go?

**Comment by liangluofb** [comment]
TODO added. The previous comments are already resolved I believe.

**Comment by min-xu-ai** [comment]
> 
> 
> TODO added. The previous comments are already resolved I believe.

Thanks! Can you see the circle CI failure in the CI pages? I think there might be some code formatting errors. Let me know if you need help fixing those.

**Comment by liangluofb** [comment]
> > TODO added. The previous comments are already resolved I believe.
> 
> Thanks! Can you see the circle CI failure in the CI pages? I think there might be some code formatting errors. Let me know if you need help fixing those.

It would be great if you can point to me which line is causing the linter rage :) I can't find it. The error just says "it's going to be reformatted" :)

**Comment by anj-s** [comment]
> > > TODO added. The previous comments are already resolved I believe.
> > 
> > 
> > Thanks! Can you see the circle CI failure in the CI pages? I think there might be some code formatting errors. Let me know if you need help fixing those.
> 
> It would be great if you can point to me which line is causing the linter rage :) I can't find it. The error just says "it's going to be reformatted" :)

You need to run "you can run black . && isort . && flake8 && mypy --ignore-missing-imports --scripts-are-modules --pretty . && echo ALL GOOD to fix the lint errors" locally which will fix lint errors.  It will reorder the required imports which is why the linter is failing.

**Comment by liangluofb** [comment]
weird. I didnt change any of the imports. Did that and still failing.

**Comment by anj-s** [comment]
> weird. I didnt change any of the imports. Did that and still failing.

Sorry I misread the linter and thought it was isort. The command should still fix it. Does it pass locally? i.e say ALL GOOD?

**Comment by liangluofb** [comment]
It throws many many errors :( 

```
stubs/torch/nn/modules/adaptive.pyi:10: error: First argument to namedtuple() should be "_ASMoutput", not "ASMoutput"
    _ASMoutput = namedtuple('ASMoutput', ['output', 'loss'])
                 ^
fairscale/utils/reduce_scatter_bucketer.py:30: error: Module has no attribute "_reduce_scatter_base"; maybe "reduce_scatter"?
                dist._reduce_scatter_base(
                ^
fairscale/utils/reduce_scatter_bucketer.py:138: error: Module has no attribute "_reduce_scatter_base"; maybe "reduce_scatter"?
                    dist._reduce_scatter_base(output, input_flattened, group=group)
                    ^
fairscale/experimental/nn/offload.py:473: error: Argument 1 to "_split" has incompatible type "Union[Sequential, List[Any]]"; expected "Sequential"
                splits = _split(model, num_slices)
                                ^
fairscale/experimental/nn/distributed_pipeline/graph.py:198: error: "Tensor" not callable
                    remote_module = partition[0].module.get_module_rref()
                                    ^
fairscale/experimental/nn/distributed_pipeline/graph.py:203: error: "Tensor" not callable
                        args=([p.module.get_module_rref() for p in partition],),
                               ^
fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1467: error: Module has no attribute "_all_gather_base"; maybe "all_gather"?
                            dist._all_gather_base(output_tensor, p_data, group=self.process_group)
                            ^
benchmarks/experimental/experimental_async_approaches.py:27: error: Module "fairscale.nn.pipe" has no attribute "LazyModule"
```

**Comment by min-xu-ai** [comment]
@liangluofb, sorry for the late reply. I was off today. There are 5 failures on CI:

image

I don't know if you can click on "Details" above and see the detailed error msg. If you can't, here is what I see.

The first 3 are related to formatting. The error is from black, not isort.

image

Did you run `pip install -r requirements-test.txt` and install the right version of black and other tools. If you have the right tools with right version, they should run cleanly on master.

I wish I could just get your branch from this PR and fix the errors easily but I don't know how to do that.

The rest 2 errors related to gpu_tests_190 is for gpu test with pytorch 1.9. One of the error is related to all_gather. Can you run this test locally and see if the error is there?

image

The other error is this:

image


image


These look like real errors.

CC: @zhaojuanmao in case she is interested in the test failures related to all_gather here.

**Comment by anj-s** [comment]
This is what I get when I clone the liangluofb:master branch and run the [above](https://github.com/facebookresearch/fairscale/pull/729#issuecomment-881038797) linter command.
```
(fs_env) anj@devfair0443:~/testing_fairscale_mypy/fairscale$ git branch
* master
(fs_env) anj@devfair0443:~/testing_fairscale_mypy/fairscale$ fs_test
reformatted /private/home/anj/testing_fairscale_mypy/fairscale/fairscale/utils/reduce_scatter_bucketer.py
reformatted /private/home/anj/testing_fairscale_mypy/fairscale/tests/nn/data_parallel/test_fsdp_overlap.py
reformatted /private/home/anj/testing_fairscale_mypy/fairscale/fairscale/nn/data_parallel/fully_sharded_data_parallel.py
All done! ✨ 🍰 ✨
3 files reformatted, 191 files left unchanged.
Skipped 2 files
fairscale/optim/adascale.py:536: error: Call to untyped function "append" in typed context
                self._state[name] = np.append(self._state[name], val)
                                    ^
Found 1 error in 1 file (checked 190 source files)
```
I did not find any test failures that Min mentioned above, maybe it was fixed in the last commit. The last lint error regarding np.append has been fixed recently. Have you merged the latest fairscale/main branch changes? 

[my conda env](https://gist.github.com/anj-s/39b524d7e6a84534a5b37f21fa33c6ef)

**Comment by min-xu-ai** [comment]
> I did not find any test failures that Min mentioned above, maybe it was fixed in the last commit.

Thanks, @anj-s! I saw the error from CI pages, BTW.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1467]:
```suggestion
                        # New version of PyTorch has all_gather_base, which is faster than chunk and then all_gather.
                        dist._all_gather_base(output_tensor, p_data, group=self.process_group)  # type: ignore
```

Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1467]: For "# type: ignore", can you add the _all_gather_base typing to the stubs dir?

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (min-xu-ai) [inline fairscale/utils/reduce_scatter_bucketer.py:30]: remove this line?

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (min-xu-ai) [inline fairscale/utils/reduce_scatter_bucketer.py:31]: Save as above, can you add the typing info to the stubs dir? If you prefer me to add that's totally fine. I can do them in a follow up PR.

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (min-xu-ai) [inline fairscale/utils/reduce_scatter_bucketer.py:131]:
```suggestion
            # TODO: investigate how to avoid using torch.cat (because it seems to be slow for CPU tensors)
```

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (min-xu-ai) [inline fairscale/utils/reduce_scatter_bucketer.py:135]: remove this line?

**Review by min-xu-ai — APPROVED** [review]
Really nice now. Thanks so much for updating the tests and address the failures! Really awesome work! If would be nice to address those comments before merging but I can do those changes in a separate PR later if you prefer.

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (anj-s) [reply fairscale/utils/reduce_scatter_bucketer.py:131]: I don't think this was about CPU tensors. torch.cat results in 1 tensor copy which we can avoid if we manipulate the tensors in place.

**Review by anj-s — APPROVED** [review]

File: tests/nn/data_parallel/test_fsdp_overlap.py
Comment (anj-s) [inline tests/nn/data_parallel/test_fsdp_overlap.py:28]: Where is this used?

File: fairscale/utils/reduce_scatter_bucketer.py
Comment (liangluofb) [reply fairscale/utils/reduce_scatter_bucketer.py:30]: Thanks! That would be great!

File: tests/nn/data_parallel/test_fsdp_overlap.py
Comment (liangluofb) [reply tests/nn/data_parallel/test_fsdp_overlap.py]: removed!

**Pull request merged**
